### PR TITLE
Deck billing status groundwork: derived properties, count fixes, status report (#1729 PR 1)

### PR DESCRIPTION
**Pull request by tylerecouture**
### What?
The preparation + step 1 of the #1729 epic (automated payments & onboarding):

* **`PLAN-1729-automated-payments-onboarding.md`** — the epic's implementation plan (phasing, architecture decisions, open questions), included here so the roadmap merges alongside its first step.
* **Billing constants** on the tenant app: `TRIAL_MAX_ACTIVE_USERS = 5` and `GRACE_PERIOD_DAYS = 30` (codifying the 0–30-day "gold band" the tenant admin changelist has shown since #1494).
* **Derived status properties on `Tenant`** — `subscription_active`, `in_grace_period`, `is_on_trial`, `is_suspended`, `effective_max_active_users`, `days_until_expiry` — computed from the existing `trial_end_date`/`paid_until` fields. No stored status field, no migration.
* **Bug fix:** `get_active_user_count()` double-counted staff who are also enrolled in a course; it now passes `students_only=True`, which also stops test accounts counting toward a deck's cap.
* **Hardening:** `update_cached_fields()` saves with `update_fields=[...]` so a stale instance can't clobber a concurrent admin edit (e.g. to `paid_until`).
* **New read-only `deck_status_report` management command** printing each deck's derived status and stored-vs-recomputed active-user counts.

### Why?
Everything later in the epic (trial banner, cap enforcement, reminder cadence, Stripe sync, suspension) reads deck status from one place. Deriving it from the dates admins already see and edit means status can never drift and no production data migration is needed. The count fix must land first because those counts become enforced caps later — and since fixing it changes real numbers for every deck, the report command exists so we can review per-deck deltas in production **before** any enforcement or notification behavior ships (plan §10.1–10.2).

### How?
Pure model layer + one management command; zero behavior change for existing decks (nothing reads the new properties yet). Decks with **both** dates blank are treated as comped/unmanaged and never suspend — that state is only reachable by explicit admin edit, and is the escape hatch for legacy/invoiced decks. `-1` (unlimited) is passed through in every state.

### Testing?
* 20 new tests: frozen-clock (freezegun) property tests covering every state boundary (grace edges, cleared-trial-with-lapsed-paid edge case, unlimited passthrough), DB-backed count tests, a concurrent-edit clobber test, and command tests.
* Test-driven per convention: the staff double-count test **fails without the fix**.
* Full suite serial run: **1200 tests, OK**; `ruff`, `manage.py check`, and `makemigrations --check --dry-run` all clean.

### Screenshots (if front end is affected)
No front-end changes. Sample report output:
```
deck                     status     trial_end    paid_until   cached  fresh  delta
----------------------------------------------------------------------------------
test                     trial      2026-09-18   -                 0      2     +2
----------------------------------------------------------------------------------
1 deck(s); 1 with a count delta. No changes were written.
```

### Anything Else?
* Next steps per the plan's phasing: PR 2 (nightly cached-field refresh, removing the `TenantAdmin.get_queryset` N+1) and PR 3 (status banner).
* The plan's §11 lists five open questions (suspension semantics for #1734, grace length, cadence, tiers, trial length) — answers there shape PRs 4+ but nothing in this PR depends on them.

### Review request
@tylerecouture

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01UAo12812TYt3GJgpZi7Pmn

- Claude Code

---
_Generated by [Claude Code](https://claude.ai/code/session_01UAo12812TYt3GJgpZi7Pmn)_

## Summary by CodeRabbit

* **New Features**
  * Added derived billing status indicators for trials, grace periods, subscriptions, and suspensions.
  * Added a read-only command to report billing status and compare cached versus current active-user counts.

* **Bug Fixes**
  * Improved active-user counting to avoid duplicate staff counts and exclude test accounts.
  * Prevented cached updates from overwriting unrelated changes.

* **Documentation**
  * Added a detailed implementation plan for automated Stripe payments, onboarding, notifications, enforcement, and rollout.

### CODE REVIEW

**Comment by tylerecouture** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by tylerecouture** [comment]
@coderabbitai review

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by tylerecouture — COMMENTED** [review]
@claude conduct your OWN full review of this PR, to ensure we are getting this feature started corectly.

**Comment by tylerecouture** [comment]
## Self-review of this PR (as requested)

I ran a full multi-lens review of the diff — logic correctness, project conventions & test quality, multitenancy & performance, and security/ops/product-fit — and independently fact-checked every candidate finding against the actual code before accepting it: **9 distinct findings confirmed** (fixed in b279589 where clear-cut), **7 candidates refuted** on verification.

### Verified sound (no action)
- Property boundary logic: inclusive at `paid_until` and at `paid_until + 30d` grace, trial edge, and the full suspension truth table (including the cleared-trial + lapsed-paid edge case) all behave as the plan specifies.
- The report command is genuinely read-only and production-safe mid-day: short per-schema SELECT/COUNTs, no writes, no locks beyond ACCESS SHARE, and it prints no PII (schema names, dates, counts only — owner name/email deliberately omitted).
- Schema correctness: counts run inside `tenant_context` against the right schema; `SiteConfig.get()`'s cache is keyed per schema (no cross-tenant bleed); the new properties are pure date arithmetic with no ORM access, so no search-path hazard when later called from request contexts.
- Conventions: the new tests pass the `test_conventions.py` naming/docstring guard, ruff is clean, `SimpleTestCase` for the pure-date property tests is appropriate, and the count bug fix is test-driven (fails without the fix).

### Findings — fixed in b279589
1. **`days_until_expiry` used the ancient trial date for lapsed subscribers** (medium). `trial_end_date` is set at creation and never cleared on subscribing, so a lapsed ex-subscriber would have reported expiry relative to a years-old trial date (with a discontinuous jump when grace ended). Suspended decks now use the **latest** lapsed clock; new test pins it.
2. **`date.today()` vs `TIME_ZONE`** (medium). The properties keyed "today" off the container's OS clock (UTC in deployment) while the admin changelist compares against `{% now %}` in America/Vancouver — decks would flip state 7–8 h early each evening relative to what admins see. Now `timezone.localdate()`. (`default_trial_end_date` still uses `date.today()` — pre-existing, follow-up candidate.)
3. **Report deltas are destroyed by any admin changelist visit** (medium). The changelist's refresh loop rewrites cached counts with the new formula, flattening the before/after evidence the command exists to show. Sequencing caveat now documented in the command help/docstring and plan §10.1: run the report **before** anyone opens the tenant changelist post-deploy.
4. **Shared-library schema appeared in the report** as a bogus trial/suspended "deck" (medium/low). Now excluded (per the plan's own rule that lifecycle machinery must skip public **and** library), with a test.
5. **One broken deck aborted the whole report** (low). A schema missing its `SiteConfig` (or the schema itself) raised through `handle()`. Each deck's recompute is now savepoint-wrapped; failures print an `ERROR` line and the report continues, with a test.
6. **Vacuous test assertion** (low): `assertIn('trial', output)` always passed because the header contains `trial_end`. Now asserts the status on the deck's own report line.
7. **Docstring overclaims** (low/nit): the test-account exclusion only applies to the enrolled-student term, and the comped "both dates blank" state isn't actually reachable through the admin form — both corrected to say what the code does.
8. **Plan/housekeeping** (nit): the plan referenced a `--dry-run` flag the command doesn't have (it's unconditionally read-only) — wording fixed; the dated plan moved from the repo root to `docs/plans/`; the count-test class renamed/redocumented to cover its caching test.

### Two product decisions for @tylerecouture (flagged, not blocking)
- **Who consumes a seat in the staff term of `active_user_count`?** Staff flagged as test accounts, deactivated staff, and the ByteDeck-owned deck admin all currently count. This becomes billing-relevant when caps are enforced (plan PR 4) — happy to filter the staff term in a follow-up if you want the stricter definition.
- **Comped decks**: reaching the "both dates blank / never suspend" state requires a shell edit today because `trial_end_date` is `null=True` but not `blank=True`. Adding `blank=True` is a one-line change plus a trivial migration — worth folding into PR 2?

### Notable refuted candidates (checked, deliberately unchanged)
- *"`effective_max_active_users` can raise an admin-set cap below 5 for trial decks"* — substitution (not clamping) is the plan's documented intent, and nothing consumes the property yet.
- *"Day-0 off-by-one vs the admin gold band"* — the JS band colors the expiry day gold while the property still counts it as subscribed; a real 1-day nuance, but display-only, and treating the last paid day as paid is the internally consistent choice.
- *"`update_fields` turns a concurrent-delete race into a DatabaseError"* — mechanically true but strictly narrower than the previous full-row clobber it replaces.

**Verification of the fix commit:** full serial suite **1204 tests OK**; ruff, `manage.py check`, `makemigrations --check --dry-run` all clean; patch coverage stays at 100% (all new branches tested).

**Verdict:** with b279589, I believe this groundwork is correct, matches the epic plan, and changes no behavior for existing decks; the two flagged items are product decisions that shape PRs 2 and 4 rather than defects here.

- Claude Code

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/tenant/models.py [hunk 1]
```diff
@@ -150,7 +244,15 @@ def update_cached_fields(self):
...
-        self.save()
+        self.save(update_fields=[
+            'owner_full_name_cached',
+            'owner_email_cached',
+            'active_user_count',
+            'total_user_count',
+            'quest_count',
+            'last_staff_login',
+            'google_signon_enabled',
+        ])
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/tenant/models.py:254]: Declining this one deliberately, per the repo convention (CLAUDE.md): *"Prefer calling `model.full_clean()` before `model.save()` in new code (much existing code doesn't — match the surrounding style rather than adding drive-by `full_clean()` calls)."*

- `update_cached_fields()` is pre-existing code (this PR only narrowed its save with `update_fields`), and it runs in bulk contexts — the admin changelist refresh loop today, a nightly beat task later (#1729 plan PR 2). Adding `full_clean()` there would run unique-checks plus every field validator on each refresh, and a single legacy row that fails a *current* validator would break the whole admin changelist / nightly refresh for that deck. A cached-counter refresh writing derived integers isn't where validation belongs.
- The test's `profile.save()` flips one boolean on a baker-built profile; `full_clean()` there could fail on unrelated randomized fields and matches no surrounding test style.

- Claude Code

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/tenant/models.py [hunk 2]
```diff
@@ -193,11 +295,20 @@ def get_total_user_count(self):
...
+        enrolled), and so enrolled (non-staff) test accounts don't count toward a
+        deck's active-user cap. Note the staff term is unconditional: staff flagged
+        as test accounts, deactivated staff, and the ByteDeck-owned deck admin all
+        still count -- whether they should is an open question for the enforcement
+        work (#1729).
         """
         staff_count = User.objects.filter(is_staff=True).count()
         CourseStudent = apps.get_model('courses', 'CourseStudent')
-        active_student_count = CourseStudent.objects.all_users_for_active_semester().count()
+        active_student_count = CourseStudent.objects.all_users_for_active_semester(students_only=True).count()
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/tenant/models.py]: Addressed in e254b28 by a stronger rule than suggested: per @tylerecouture's decision, **staff and superusers now don't count at all** — `get_active_user_count()` counts only students registered in a course in the active semester (`students_only=True`, which also excludes test accounts and archived users; superusers excluded explicitly). The cached field's help_text and the plan's "active user" definition are updated to match (migration 0016, DB no-op), and the count tests are rewritten around the students-only definition.

- Claude Code

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/tenant/tests/test_management_commands.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        self.assertEqual(len(matches), 1, f"expected exactly one report line for {schema_name}:\n{output}")
+        return matches[0]
+
+    def make_schemaless_tenant(self, name):
+        """Create a Tenant row without creating its Postgres schema (fast, and lets tests
+        exercise rows the command must skip or report as errors)."""
+        with schema_context(get_public_schema_name()):
+            tenant = Tenant(name=name, schema_name=name)
+            tenant.auto_create_schema = False
+            tenant.save()
```
[the remark this answers is not in the document]
Comment (tylerecouture) [reply src/tenant/tests/test_management_commands.py:43]: Done in e254b28 — the fixture helper now calls `tenant.full_clean()` before saving, matching the repo's prefer-full_clean-in-new-code convention.

- Claude Code

---
_Generated by [Claude Code](https://claude.ai/code)_
